pulumi/infra/service_queue.py

```diff
@@ -83,6 +84,140 @@ def redrive_policy(arn: pulumi.Output) -> str:
+        self.retry_send_policy = _queue_send_policy(
+            logical_retry_name, self.retry_queue
+        )
+        self.dead_letter_send_policy = _queue_send_policy(
+            logical_dead_letter_name, self.dead_letter_queue
+        )
 
         self.register_outputs({})
+
+    # Yes, the `id` property of an SQS queue is actually a URL.
+    #
+    # The URL generated by Pulumi is currently of the form:
+    #     https://sqs.{AWS_REGION}.amazonaws.com/{ACCOUNT_ID}/{QUEUE_NAME}
+    #
+    # FYI: A URL in the form of:
+    #     https://queue.amazonaws.com/{ACCOUNT_ID}/{QUEUE_NAME}
+    # would also be valid, but this is not what Pulumi generates.
+    #
+    # We expose these URLs as properties on ServiceQueue to:
+    #
+    # a) encapsulate things a bit better
+    # b) only have one place that needs to know that "id == URL"
+    # c) have one place to modify if this behavior ever changes in Pulumi
+
+    @property
+    def main_queue_url(self) -> pulumi.Output[str]:
+        return self.queue.id
+
+    @property
+    def retry_queue_url(self) -> pulumi.Output[str]:
+        return self.retry_queue.id
+
+    @property
+    def dead_letter_queue_url(self) -> pulumi.Output[str]:
+        return self.dead_letter_queue.id
+
+    def grant_main_queue_consumption_to(self, role: aws.iam.Role) -> None:
+        attach_policy(self.queue_consumption_policy, role)
+
+    def grant_retry_queue_consumption_to(self, role: aws.iam.Role) -> None:
+        attach_policy(self.retry_consumption_policy, role)
+
+    def grant_dead_letter_queue_consumption_to(self, role: aws.iam.Role) -> None:
+        attach_policy(self.dead_letter_consumption_policy, role)
+
+    def grant_main_queue_send_to(self, role: aws.iam.Role) -> None:
+        attach_policy(self.queue_send_policy, role)
+
+    def grant_retry_queue_send_to(self, role: aws.iam.Role) -> None:
+        attach_policy(self.retry_send_policy, role)
+
+    def grant_dead_letter_queue_send_to(self, role: aws.iam.Role) -> None:
+        attach_policy(self.dead_letter_send_policy, role)
+
+    def subscribe_to_emitter(self, emitter: EventEmitter) -> None:
+        """
+        Enable this queue to be fed by events from `emitter`.
+        """
+        aws.sns.TopicSubscription(
+            f"{self.queue._name}-subscribes-to-{emitter.topic._name}",
+            protocol="sqs",
+            endpoint=self.queue.arn,
+            topic=emitter.topic.arn,
+            raw_message_delivery=True,
+            opts=pulumi.ResourceOptions(parent=emitter.topic),
+        )
+
+
+def _queue_consumption_policy(queue_name: str, queue: aws.sqs.Queue) -> aws.iam.Policy:
+    """Create an IAM Policy to consume messages from the given SQS queue.
+
+    queue_name is folded into the Policy resource name, and has to be a string (rather than a Pulumi Output).
+    """
+    return aws.iam.Policy(
+        f"consume-from-{queue_name}",
+        # We interpolate on queue.name to get the physical name
```

Comment:
(we probably will never have a situation where many things pull from same queue, right?)

Reply:
Probably not... I don't think this code is relevant to that, though.